File: Scala/src/main/scala/TextbookTrack/Chapter09/BA9B.scala
package TextbookTrack.Chapter09

object BA9B {
  import scala.annotation.tailrec
  import scala.collection.mutable.{Map => MutableMap}
  import scala.collection.{Map => BothMap}

  private type Node = Int

  private val StringTerminator: Char = '$'

  final case class Edge(nodeFrom: Node, nodeTo: Node, label: Char) {
    override def toString: String = s"$nodeFrom->$nodeTo:$label"
  }

  class Trie(patterns: List[String]) {
    import Trie.getNeighbourWithGivenLabel

    private val nodeIdGenerator: Iterator[Int] = Iterator.from(0)
    private val root: Int = nodeIdGenerator.next()
    private val adjacencyList: Map[Node, List[Edge]] = buildTrieFromPatterns()

    def edges(): Iterator[Edge] = adjacencyList.valuesIterator.flatten

    def isMatching(text: List[Char]): Boolean = {
      @tailrec
      def loop(xs: List[Char], currentNode: Node): Boolean = xs match {
        case Nil => false
        case letter :: xss =>
          getNeighbourWithGivenLabel(adjacencyList, currentNode, letter) match {
            case None => false
            case Some(nextNode) => if (isPatternEnd(nextNode)) true else loop(xss, nextNode)
          }
      }

      loop(text, root)
    }

    private def isPatternEnd(node: Node): Boolean =
      adjacencyList.getOrElse(node, Nil).exists{ case Edge(_, _, label) => label == StringTerminator }

    private def buildTrieFromPatterns(): Map[Node, List[Edge]] = {
      val adjacencyList: MutableMap[Node, List[Edge]] = MutableMap()
      patterns.foreach(addPatternToTrie(adjacencyList, _))
      adjacencyList.toMap
    }

    private def addPatternToTrie(adjacencyList: MutableMap[Node, List[Edge]], pattern: String): Unit = {
      @tailrec
      def loop(letters: List[Char], currentNode: Node): Unit = letters match {
        case Nil => ()
        case letter :: lss =>
          getNeighbourWithGivenLabel(adjacencyList, currentNode, letter) match {
            case None =>
              val nextNode: Node = nodeIdGenerator.next()
              val neighbours: List[Edge] = adjacencyList.getOrElse(currentNode, Nil)
              adjacencyList.update(currentNode, Edge(currentNode, nextNode, letter) :: neighbours)
              loop(lss, nextNode)
            case Some(node) => loop(lss, node)
          }
      }

      loop(pattern.toList, root)
    }
  }

  object Trie {
    private def getNeighbourWithGivenLabel(adjacencyList: BothMap[Node, List[Edge]],
                                           currentNode: Node,
                                           letter: Char): Option[Node] = {
      adjacencyList
        .get(currentNode)
        .flatMap(_.find{ case Edge(_, _, label) => label == letter })
        .map(_.nodeTo)
    }
  }

  def runTrieMatching(text: String, patterns: List[String]): List[Int] = {
    val trie = new Trie(patterns.map(_ + StringTerminator))
    @tailrec
    def loop(matchingIndices: List[Int], text: List[Char], ix: Int): List[Int] = text match {
      case Nil => matchingIndices.reverse
      case _ :: rest =>
        if (trie.isMatching(text)) loop(ix :: matchingIndices, rest, ix + 1)
        else loop(matchingIndices, rest, ix + 1)
    }

    loop(Nil, text.toList, 0)
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val text: String = reader.next()
    val patterns: List[String] = reader.toList
    val result: List[Int] = runTrieMatching(text, patterns)
    println(result.mkString(" "))
  }
}

File: Python/textbook_track/chapter09/ba9b.py
# Implement TrieMatching
import itertools as it
import sys

from collections import defaultdict


class Trie:
    def __init__(self, patterns):
        self._node_id_generator = it.count()
        self._root = next(self._node_id_generator)
        self._adjacency_list = self._build_trie_from_patterns(patterns)

    def edges(self):
        for node, neighbours in self._adjacency_list.items():
            for neighbour, label in neighbours:
                yield f'{node}->{neighbour}:{label}'

    def is_matching(self, text):
        current_node = self._root
        for letter in text:
            next_node = _get_neighbour_with_given_label(self._adjacency_list, current_node, letter)
            if next_node is None:
                return False

            if self._is_pattern_end(next_node):
                return True

            current_node = next_node

        return False

    def _build_trie_from_patterns(self, patterns):
        adjacency_list = defaultdict(list)
        for pattern in patterns:
            self._add_pattern_to_trie(adjacency_list, pattern)
        return dict(adjacency_list)

    def _add_pattern_to_trie(self, adjacency_list, pattern):
        current_node = self._root
        for letter in pattern:
            node = _get_neighbour_with_given_label(adjacency_list, current_node, letter)
            if node is None:
                next_node = next(self._node_id_generator)
                adjacency_list[current_node].append((next_node, letter))
                current_node = next_node
            else:
                current_node = node

    def _is_pattern_end(self, node):
        neighbours = self._adjacency_list.get(node, [])
        return any(label == '$' for _, label in neighbours)


def _get_neighbour_with_given_label(adjacency_list, current_node, letter):
    neighbours = adjacency_list.get(current_node, [])
    for node, label in neighbours:
        if label == letter:
            return node
    return None


def run_trie_matching(text, patterns):
    trie = Trie(patterns)
    matching_indices = []
    for ix in range(len(text)):
        if trie.is_matching(text[ix:]):
            matching_indices.append(ix)
    return matching_indices


def main():
    data = sys.stdin.read().splitlines()
    text = data[0]
    patterns = map(lambda x: x + '$', data[1:])
    result = run_trie_matching(text, patterns)
    print(' '.join(map(str, result)))


if __name__ == '__main__':
    main()
